File: Code/Brain_Kuramoto/Brain_circadian.m
clear; clc; close all;

baseDir = fileparts(mfilename('fullpath'));
figDir = fullfile(baseDir, '..', '..', 'Report', 'Figs');
if ~exist(figDir, 'dir'), mkdir(figDir); end

make_fig15(figDir);
make_fig16(figDir);
make_fig17(figDir);

fprintf('Brain_circadian.m finished: Fig.15-17 are generated.\n');

function make_fig15(figDir)
delta = 1.0;
f = 3.5 * delta;
omega = 1.4 * delta;
regimes = [10.0, 4.5] * delta;
regimeLabels = {'$K=10\Delta$', '$K=4.5\Delta$'};
shifts = [-8.5, -9.5, 9.0, 12.0];
cols = twilight_colors(numel(shifts));
fig = figure('Color', 'w', 'Position', [100 100 720 335]);
tiledlayout(1, 2, 'TileSpacing', 'compact', 'Padding', 'compact');
thetaCircle = linspace(0, 2*pi, 500);
for r = 1:2
    ax = nexttile; hold on;
    k = regimes(r);
    zst = stable_point_fast(k, f, delta, omega);
    pts = fixed_points_all(k, f, delta, omega);
    plot(cos(thetaCircle), sin(thetaCircle), '--', 'LineWidth', 1.0, 'Color', [0.82 0.84 0.86]);
    plot([-1 1], [0 0], 'Color', [0.82 0.84 0.86], 'LineWidth', 0.9);
    plot([0 0], [-1 1], 'Color', [0.82 0.84 0.86], 'LineWidth', 0.9);
    for rad = [0.25 0.55 0.85]
        for ang = linspace(0, 2*pi, 18)
            z0 = rad * exp(1i*ang);
            [~, z] = trajectory(z0, 25, k, f, delta, omega, 1200, 0.035);
            plot(real(z), imag(z), 'Color', [0.82 0.84 0.86], 'LineWidth', 0.55);
        end
    end
    for s = 1:numel(shifts)
        z0 = zst * exp(1i * shifts(s) * 2*pi/24);
        [~, z] = trajectory(z0, 80, k, f, delta, omega, 2600, 0.035);
        plot(real(z), imag(z), 'LineWidth', 1.5, 'Color', cols(s,:));
        plot(real(z(1)), imag(z(1)), 'o', 'MarkerSize', 3.2, 'MarkerFaceColor', cols(s,:), 'MarkerEdgeColor', cols(s,:));
    end
    for p = 1:size(pts,1)
        z = pts(p,1) + 1i*pts(p,2);
        ev = eig(jacobian(z, k, f, delta, omega));
        if all(real(ev) < 0)
            plot(real(z), imag(z), 'p', 'MarkerSize', 7, 'MarkerFaceColor', 'k', 'MarkerEdgeColor', 'k');
        elseif all(real(ev) > 0)
            plot(real(z), imag(z), 'o', 'MarkerSize', 5, 'MarkerFaceColor', 'w', 'MarkerEdgeColor', 'k');
        else
            plot(real(z), imag(z), '+', 'MarkerSize', 8, 'LineWidth', 1.5, 'Color', 'k');
        end
    end
    text(-0.13, 1.04, char('A'+r-1), 'Units', 'normalized', 'FontName', 'Arial', 'FontSize', 11);
    text(0.05, 0.92, regimeLabels{r}, 'Units', 'normalized', 'Interpreter', 'latex', 'FontName', 'Arial', 'FontSize', 9);
    axis equal; xlim([-1.02 1.02]); ylim([-1.02 1.02]);
    xlabel('$\operatorname{Re}(z)$', 'Interpreter', 'latex');
    if r == 1, ylabel('$\operatorname{Im}(z)$', 'Interpreter', 'latex'); end
    style_axes(ax);
end
exportgraphics(fig, fullfile(figDir, '15_circadian_phase_space_dynamics.png'), 'Resolution', 300, 'BackgroundColor', 'w');
end

function make_fig16(figDir)
delta = 3.8e-3;
k = 4.5 * delta; f = 3.5 * delta; omega = 1.4 * delta;
zst = stable_point_fast(k, f, delta, omega);
cases = [-3 -6 -9 12 3 6 9];
labels = {'3 E','6 E','9 E','12 E/W','3 W','6 W','9 W'};
styles = {'-','-','-','-','--','--','--'};
cols = twilight_colors(4);
caseCols = [cols(1,:); cols(2,:); cols(3,:); cols(4,:); cols(1,:); cols(2,:); cols(3,:)];
fig = figure('Color', 'w', 'Position', [100 100 470 335]); hold on;
for i = 1:numel(cases)
    z0 = zst * exp(1i * cases(i) * 2*pi/24);
    [t, z] = trajectory(z0, 14*24, k, f, delta, omega, 900, inf);
    d = abs(z - zst);
    stop = find(d <= 0.2, 1, 'first');
    if isempty(stop), stop = numel(d); end
    plot(t(1:stop)/24, d(1:stop), styles{i}, 'LineWidth', 1.6, 'Color', caseCols(i,:), 'DisplayName', labels{i});
end
yline(0.2, 'k-', 'LineWidth', 1.1);
xlabel('Days'); ylabel('$|z(t)-z_{st}|$', 'Interpreter', 'latex');
xlim([0 14]); ylim([0 2.0]);
legend('Location', 'northeast', 'NumColumns', 2, 'Box', 'off');
style_axes(gca);
exportgraphics(fig, fullfile(figDir, '16_circadian_recovery_trajectories.png'), 'Resolution', 300, 'BackgroundColor', 'w');
end

function make_fig17(figDir)
delta = 3.8e-3;
kRef = 4.5 * delta; fRef = 3.5 * delta; omegaRef = 1.4 * delta;
cases = [-3 -6 -9 12 3 6 9];
labels = {'3 E','6 E','9 E','12 E/W','3 W','6 W','9 W'};
styles = {'-','-','-','-','--','--','--'};
cols = twilight_colors(4);
caseCols = [cols(1,:); cols(2,:); cols(3,:); cols(4,:); cols(1,:); cols(2,:); cols(3,:)];
scans = {
    '$K/\Delta$', linspace(2.0, 15.0, 72), [2 24], 1;
    '$F/\Delta$', linspace(1.5, 5.8, 70), [0 45], 2;
    '$\Omega/\Delta$', linspace(-3.7, 3.7, 78), [0 34], 3
};
fig = figure('Color', 'w', 'Position', [100 100 820 280]);
tiledlayout(1, 3, 'TileSpacing', 'compact', 'Padding', 'compact');
for s = 1:3
    ax = nexttile; hold on;
    xs = scans{s,2};
    for c = 1:numel(cases)
        rec = nan(size(xs));
        for i = 1:numel(xs)
            if scans{s,4} == 1
                k = xs(i)*delta; f = fRef; omega = omegaRef;
            elseif scans{s,4} == 2
                k = kRef; f = xs(i)*delta; omega = omegaRef;
            else
                k = kRef; f = fRef; omega = xs(i)*delta;
            end
            zst = stable_point_fast(k, f, delta, omega);
            rec(i) = recovery_time(zst, cases(c), k, f, delta, omega, 0.2, 28);
        end
        plot(xs, rec, styles{c}, 'LineWidth', 1.25, 'Color', caseCols(c,:), 'DisplayName', labels{c});
    end
    text(-0.15, 1.04, char('A'+s-1), 'Units', 'normalized', 'FontName', 'Arial', 'FontSize', 11);
    xlabel(scans{s,1}, 'Interpreter', 'latex');
    if s == 1, ylabel('Recovery time (days)'); end
    ylim(scans{s,3});
    if s == 2
        legend('Location', 'northeast', 'NumColumns', 2, 'Box', 'off', 'FontSize', 6.4);
    end
    style_axes(ax);
end
exportgraphics(fig, fullfile(figDir, '17_circadian_parameter_dependence.png'), 'Resolution', 300, 'BackgroundColor', 'w');
end

function dz = dzdt_complex(z, k, f, delta, omega)
dz = 0.5 * ((k*z + f) - z.^2 .* (k*conj(z) + f)) - (delta + 1i*omega) * z;
end

function dy = rhs_real(~, y, k, f, delta, omega)
z = y(1) + 1i*y(2);
dz = dzdt_complex(z, k, f, delta, omega);
dy = [real(dz); imag(dz)];
end

function z = stable_point_fast(k, f, delta, omega)
starts = [0.5 0; 0 0; -0.5 0; 0 0.5; 0 -0.5; 0.8 0.2; -0.2 0.8];
best = starts(1,:); bestVal = inf;
opts = optimset('Display', 'off', 'TolX', 1e-10, 'TolFun', 1e-12, 'MaxIter', 800);
for i = 1:size(starts,1)
    y = fminsearch(@(yy) rhs_norm(yy, k, f, delta, omega), starts(i,:), opts);
    val = rhs_norm(y, k, f, delta, omega);
    if norm(y) <= 1.02 && val < bestVal
        best = y; bestVal = val;
    end
end
z = best(1) + 1i*best(2);
end

function pts = fixed_points_all(k, f, delta, omega)
grid = linspace(-0.92, 0.92, 11);
pts = [];
opts = optimset('Display', 'off', 'TolX', 1e-10, 'TolFun', 1e-12, 'MaxIter', 600);
for x = grid
    for y = grid
        if x^2 + y^2 > 1, continue; end
        p = fminsearch(@(yy) rhs_norm(yy, k, f, delta, omega), [x y], opts);
        if norm(p) <= 1.02 && rhs_norm(p, k, f, delta, omega) < 1e-9
            if isempty(pts) || all(sqrt(sum((pts - p).^2, 2)) > 1e-4)
                pts = [pts; p]; %#ok<AGROW>
            end
        end
    end
end
if isempty(pts)
    z = stable_point_fast(k, f, delta, omega);
    pts = [real(z) imag(z)];
end
end

function val = rhs_norm(y, k, f, delta, omega)
if sum(y.^2) > 1.08
    val = 1e3 + sum(y.^2);
else
    r = rhs_real(0, y(:), k, f, delta, omega);
    val = sum(r.^2);
end
end

function J = jacobian(z, k, f, delta, omega)
epsv = 1e-5; y = [real(z); imag(z)]; J = zeros(2);
for i = 1:2
    e = zeros(2,1); e(i) = epsv;
    J(:,i) = (rhs_real(0, y+e, k, f, delta, omega) - rhs_real(0, y-e, k, f, delta, omega)) / (2*epsv);
end
end

function [t, z] = trajectory(z0, tEnd, k, f, delta, omega, n, maxStep)
if isinf(maxStep)
    opts = odeset('RelTol', 1e-8, 'AbsTol', 1e-10);
else
    opts = odeset('RelTol', 1e-8, 'AbsTol', 1e-10, 'MaxStep', maxStep);
end
tspan = linspace(0, tEnd, n);
[t, y] = ode45(@(tt, yy) rhs_real(tt, yy, k, f, delta, omega), tspan, [real(z0); imag(z0)], opts);
z = y(:,1) + 1i*y(:,2);
end

function days = recovery_time(zst, shiftHours, k, f, delta, omega, threshold, maxDays)
z0 = zst * exp(1i * shiftHours * 2*pi/24);
[t, z] = trajectory(z0, 24*maxDays, k, f, delta, omega, 850, inf);
d = abs(z - zst);
idx = find(d <= threshold, 1, 'first');
if isempty(idx), days = nan; else, days = t(idx) / 24; end
end

function style_axes(ax)
set(ax, 'FontName', 'Arial', 'FontSize', 8.5, 'LineWidth', 1.1, 'TickDir', 'out', 'Box', 'off');
grid(ax, 'off');
end

function cols = twilight_colors(n)
base = [0.8858 0.8500 0.8879; 0.6443 0.4390 0.7389; 0.1849 0.0794 0.2131; ...
        0.2453 0.2878 0.5373; 0.2890 0.5650 0.7595; 0.7590 0.7060 0.5680; ...
        0.8858 0.8500 0.8879];
x = linspace(0,1,size(base,1)); xi = linspace(0.08,0.88,n);
cols = interp1(x, base, xi);
end

File: Code/Brain_Kuramoto/Brain_circadian.py
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from scipy.integrate import solve_ivp
from scipy.optimize import minimize


BASE = Path(__file__).resolve().parent
FIG_DIR = BASE.parent.parent / "Report" / "Figs"


def twilight_colors(n):
    cmap = plt.get_cmap("twilight")
    return np.array([cmap(x)[:3] for x in np.linspace(0.08, 0.88, n)])


def style_axes(ax):
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.tick_params(direction="out", width=1.1, labelsize=8.5)
    for spine in ax.spines.values():
        spine.set_linewidth(1.1)


def dzdt_complex(z, k, f, delta, omega):
    return 0.5 * ((k * z + f) - z * z * (k * np.conj(z) + f)) - (delta + 1j * omega) * z


def rhs_real(_, y, k, f, delta, omega):
    z = y[0] + 1j * y[1]
    dz = dzdt_complex(z, k, f, delta, omega)
    return np.array([dz.real, dz.imag])


def rhs_norm(y, k, f, delta, omega):
    y = np.asarray(y)
    if np.sum(y * y) > 1.08:
        return 1e3 + float(np.sum(y * y))
    r = rhs_real(0, y, k, f, delta, omega)
    return float(r @ r)


def stable_point_fast(k, f, delta, omega):
    starts = np.array([[0.5, 0], [0, 0], [-0.5, 0], [0, 0.5], [0, -0.5], [0.8, 0.2], [-0.2, 0.8]])
    best, best_val = starts[0], np.inf
    for start in starts:
        res = minimize(lambda yy: rhs_norm(yy, k, f, delta, omega), start, method="Nelder-Mead", options={"maxiter": 800, "xatol": 1e-10, "fatol": 1e-12})
        val = rhs_norm(res.x, k, f, delta, omega)
        if np.linalg.norm(res.x) <= 1.02 and val < best_val:
            best, best_val = res.x, val
    return best[0] + 1j * best[1]


def fixed_points_all(k, f, delta, omega):
    pts = []
    grid = np.linspace(-0.92, 0.92, 11)
    for x in grid:
        for y in grid:
            if x * x + y * y > 1:
                continue
            res = minimize(lambda yy: rhs_norm(yy, k, f, delta, omega), [x, y], method="Nelder-Mead", options={"maxiter": 600, "xatol": 1e-10, "fatol": 1e-12})
            p = res.x
            if np.linalg.norm(p) <= 1.02 and rhs_norm(p, k, f, delta, omega) < 1e-9:
                if not pts or np.all(np.linalg.norm(np.asarray(pts) - p, axis=1) > 1e-4):
                    pts.append(p)
    if not pts:
        z = stable_point_fast(k, f, delta, omega)
        pts.append([z.real, z.imag])
    return np.asarray(pts)


def jacobian(z, k, f, delta, omega):
    eps = 1e-5
    y = np.array([z.real, z.imag])
    mat = np.zeros((2, 2))
    for i in range(2):
        e = np.zeros(2)
        e[i] = eps
        mat[:, i] = (rhs_real(0, y + e, k, f, delta, omega) - rhs_real(0, y - e, k, f, delta, omega)) / (2 * eps)
    return mat


def trajectory(z0, t_end, k, f, delta, omega, n, max_step=np.inf):
    kwargs = {"rtol": 1e-8, "atol": 1e-10}
    if np.isfinite(max_step):
        kwargs["max_step"] = max_step
    t_eval = np.linspace(0, t_end, n)
    sol = solve_ivp(rhs_real, (0, t_end), [z0.real, z0.imag], t_eval=t_eval, args=(k, f, delta, omega), **kwargs)
    z = sol.y[0] + 1j * sol.y[1]
    return sol.t, z


def recovery_time(zst, shift_hours, k, f, delta, omega, threshold, max_days):
    z0 = zst * np.exp(1j * shift_hours * 2 * np.pi / 24)
    t, z = trajectory(z0, 24 * max_days, k, f, delta, omega, 850)
    hit = np.flatnonzero(np.abs(z - zst) <= threshold)
    return np.nan if len(hit) == 0 else t[hit[0]] / 24


def make_fig15(save=False):
    delta = 1.0
    f, omega = 3.5 * delta, 1.4 * delta
    regimes = [10.0 * delta, 4.5 * delta]
    shifts = [-8.5, -9.5, 9.0, 12.0]
    cols = twilight_colors(len(shifts))
    fig, axes = plt.subplots(1, 2, figsize=(7.2, 3.35), constrained_layout=True)
    theta = np.linspace(0, 2 * np.pi, 500)
    for ri, (ax, k) in enumerate(zip(axes, regimes)):
        zst = stable_point_fast(k, f, delta, omega)
        pts = fixed_points_all(k, f, delta, omega)
        ax.plot(np.cos(theta), np.sin(theta), "--", lw=1.0, color=(0.82, 0.84, 0.86))
        ax.axhline(0, color=(0.82, 0.84, 0.86), lw=0.9)
        ax.axvline(0, color=(0.82, 0.84, 0.86), lw=0.9)
        for rad in [0.25, 0.55, 0.85]:
            for ang in np.linspace(0, 2 * np.pi, 18):
                _, z = trajectory(rad * np.exp(1j * ang), 25, k, f, delta, omega, 1200, 0.035)
                ax.plot(z.real, z.imag, color=(0.82, 0.84, 0.86), lw=0.55)
        for shift, col in zip(shifts, cols):
            z0 = zst * np.exp(1j * shift * 2 * np.pi / 24)
            _, z = trajectory(z0, 80, k, f, delta, omega, 2600, 0.035)
            ax.plot(z.real, z.imag, lw=1.5, color=col)
            ax.plot(z[0].real, z[0].imag, "o", ms=3.2, color=col)
        for p in pts:
            z = p[0] + 1j * p[1]
            ev = np.linalg.eigvals(jacobian(z, k, f, delta, omega))
            if np.all(ev.real < 0):
                ax.plot(z.real, z.imag, "p", ms=7, color="k")
            elif np.all(ev.real > 0):
                ax.plot(z.real, z.imag, "o", ms=5, markerfacecolor="w", markeredgecolor="k")
            else:
                ax.plot(z.real, z.imag, "+", ms=8, mew=1.5, color="k")
        ax.text(-0.13, 1.04, chr(ord("A") + ri), transform=ax.transAxes, fontname="Arial", fontsize=11)
        ax.text(0.05, 0.92, rf"$K={10 if ri == 0 else 4.5}\Delta$", transform=ax.transAxes, fontsize=9)
        ax.set_aspect("equal")
        ax.set(xlim=(-1.02, 1.02), ylim=(-1.02, 1.02), xlabel=r"$\operatorname{Re}(z)$")
        if ri == 0:
            ax.set_ylabel(r"$\operatorname{Im}(z)$")
        style_axes(ax)
    if save:
        fig.savefig(FIG_DIR / "15_circadian_phase_space_dynamics.png", dpi=300, bbox_inches="tight")
    return fig


def make_fig16(save=False):
    delta = 3.8e-3
    k, f, omega = 4.5 * delta, 3.5 * delta, 1.4 * delta
    zst = stable_point_fast(k, f, delta, omega)
    cases = [-3, -6, -9, 12, 3, 6, 9]
    labels = ["3 E", "6 E", "9 E", "12 E/W", "3 W", "6 W", "9 W"]
    styles = ["-", "-", "-", "-", "--", "--", "--"]
    base = twilight_colors(4)
    cols = [base[0], base[1], base[2], base[3], base[0], base[1], base[2]]
    fig, ax = plt.subplots(figsize=(4.7, 3.35))
    for case, label, ls, col in zip(cases, labels, styles, cols):
        t, z = trajectory(zst * np.exp(1j * case * 2 * np.pi / 24), 14 * 24, k, f, delta, omega, 900)
        d = np.abs(z - zst)
        stop = np.flatnonzero(d <= 0.2)
        end = stop[0] + 1 if len(stop) else len(d)
        ax.plot(t[:end] / 24, d[:end], ls, lw=1.6, color=col, label=label)
    ax.axhline(0.2, color="k", lw=1.1)
    ax.set(xlabel="Days", ylabel=r"$|z(t)-z_{st}|$", xlim=(0, 14), ylim=(0, 2.0))
    ax.legend(loc="upper right", ncol=2, frameon=False)
    style_axes(ax)
    if save:
        fig.savefig(FIG_DIR / "16_circadian_recovery_trajectories.png", dpi=300, bbox_inches="tight")
    return fig


def make_fig17(save=False):
    delta = 3.8e-3
    k_ref, f_ref, omega_ref = 4.5 * delta, 3.5 * delta, 1.4 * delta
    cases = [-3, -6, -9, 12, 3, 6, 9]
    labels = ["3 E", "6 E", "9 E", "12 E/W", "3 W", "6 W", "9 W"]
    styles = ["-", "-", "-", "-", "--", "--", "--"]
    base = twilight_colors(4)
    cols = [base[0], base[1], base[2], base[3], base[0], base[1], base[2]]
    scans = [
        (r"$K/\Delta$", np.linspace(2.0, 15.0, 72), (2, 24), 1),
        (r"$F/\Delta$", np.linspace(1.5, 5.8, 70), (0, 45), 2),
        (r"$\Omega/\Delta$", np.linspace(-3.7, 3.7, 78), (0, 34), 3),
    ]
    fig, axes = plt.subplots(1, 3, figsize=(8.2, 2.8), constrained_layout=True)
    for si, (ax, (xlabel, xs, ylim, mode)) in enumerate(zip(axes, scans)):
        for case, label, ls, col in zip(cases, labels, styles, cols):
            rec = []
            for x in xs:
                if mode == 1:
                    k, f, omega = x * delta, f_ref, omega_ref
                elif mode == 2:
                    k, f, omega = k_ref, x * delta, omega_ref
                else:
                    k, f, omega = k_ref, f_ref, x * delta
                zst = stable_point_fast(k, f, delta, omega)
                rec.append(recovery_time(zst, case, k, f, delta, omega, 0.2, 28))
            ax.plot(xs, rec, ls, lw=1.25, color=col, label=label)
        ax.text(-0.15, 1.04, chr(ord("A") + si), transform=ax.transAxes, fontname="Arial", fontsize=11)
        ax.set(xlabel=xlabel, ylim=ylim)
        if si == 0:
            ax.set_ylabel("Recovery time (days)")
        if si == 1:
            ax.legend(loc="upper right", ncol=2, frameon=False, fontsize=6.4)
        style_axes(ax)
    if save:
        fig.savefig(FIG_DIR / "17_circadian_parameter_dependence.png", dpi=300, bbox_inches="tight")
    return fig


def main(save=False):
    make_fig15(save)
    make_fig16(save)
    make_fig17(save)
    plt.show()


if __name__ == "__main__":
    main(save=False)
